File: python/flink_agents/runtime/skill/skill_manager.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
import contextlib
import logging
from pathlib import Path
from typing import TYPE_CHECKING, Dict, List

from flink_agents.api.skills import Skills, SkillSourceSpec
from flink_agents.runtime.skill import skill_source_registry
from flink_agents.runtime.skill.agent_skill import AgentSkill, SkillOrigin
from flink_agents.runtime.skill.skill_prompt_provider import SkillPromptProvider

if TYPE_CHECKING:
    from flink_agents.runtime.skill.skill_repository import SkillRepository

logger = logging.getLogger(__name__)


def _origin_of(spec: SkillSourceSpec) -> SkillOrigin:
    """Build a SkillOrigin from a spec for diagnostics.

    Delegates location description to the handler so a new scheme is one
    ``register()`` call, not a parallel switch here.
    """
    handler = skill_source_registry.get(spec.scheme)
    return SkillOrigin(
        scheme=spec.scheme,
        location=handler.describe_location(spec.params),
    )


class SkillManager:
    """Internal runtime component for loading, parsing, and managing skills.

    Created by the runtime from a :class:`Skills` configuration resource.
    Never exposed to users directly.

    Progressive Disclosure:
    - Discovery: Load only name/description at startup (~100 tokens)
    - Activation: Load full SKILL.md when skill matches task
    - Execution: Load resources/scripts only when needed
    """

    def __init__(self, skills_config: Skills) -> None:
        """Initialize the SkillManager from a Skills configuration."""
        self._skills: Dict[str, AgentSkill] = {}
        self._repos: Dict[str, SkillRepository] = {}
        # Every opened repo in load order, kept separately from `_repos` because that
        # map is keyed by skill name — duplicate names overwrite the earlier repo's
        # reference, so close() iterates this list (id-deduped) instead.
        self._opened_repos: List[SkillRepository] = []
        self._config = skills_config
        self._load_skills()

    @property
    def size(self) -> int:
        """Get the number of registered skills."""
        return len(self._skills)

    def get_skill(self, name: str) -> AgentSkill:
        """Get a registered skill by name."""
        if name not in self._skills:
            msg = f"Skill {name} not found, available skill names are: {list(self._skills.keys())}"
            raise ValueError(msg)
        return self._skills[name]

    def get_all_skill_names(self) -> List[str]:
        """Get the names of all registered skills."""
        return list(self._skills.keys())

    def load_skill_resource(self, skill_name: str, resource_path: str) -> str | None:
        """Load a specified resource of a skill."""
        skill = self.get_skill(skill_name)
        return skill.get_resource(resource_path)

    def generate_discovery_prompt(self, *names: str) -> str:
        """Generate a system prompt for skill discovery."""
        if self.size == 0:
            return ""

        skill_list = []
        for name in names:
            skill = self.get_skill(name)
            skill_list.append(
                SkillPromptProvider.AVAILABLE_SKILL_TEMPLATE.format(
                    name=skill.name, description=skill.description
                )
            )

        return (
            SkillPromptProvider.SKILL_DISCOVERY_PROMPT.format()
            + ("".join(skill_list))
            + SkillPromptProvider.AVAILABLE_SKILLS_TAG_END
        )

    def get_skill_dirs(self, *names: str) -> List[str]:
        """Return absolute directory paths for the given skill names.

        If no names are provided, returns directories for all filesystem-backed
        skills. Skills not backed by a filesystem repo are silently skipped.

        Raises:
            ValueError: If any provided name is not a registered skill.
        """
        selected = names if names else tuple(self._repos.keys())
        dirs: List[str] = []
        for skill_name in selected:
            repo = self._repos.get(skill_name)
            if repo is None:
                msg = (
                    f"Skill {skill_name} not found, "
                    f"available skill names are: {list(self._repos.keys())}"
                )
                raise ValueError(msg)
            dir_path = repo.get_skill_dir(skill_name)
            if dir_path is not None:
                dirs.append(str(dir_path))
        return dirs

    def get_skill_dir(self, skill_name: str) -> Path | None:
        """Return absolute directory path for a single skill, if filesystem-backed."""
        repo = self._repos.get(skill_name)
        return None if repo is None else repo.get_skill_dir(skill_name)

    def _load_skills(self) -> None:
        try:
            for spec in self._config.sources:
                try:
                    handler = skill_source_registry.get(spec.scheme)
                    repo = handler.open(spec.params)
                    self._opened_repos.append(repo)
                except (OSError, ValueError) as e:
                    msg = f"Failed to load skills from {spec.scheme}:{spec.params}"
                    raise RuntimeError(msg) from e
                self._register_repo(repo, _origin_of(spec))
        except BaseException:
            # Release every repo opened so far — the caller never receives a
            # SkillManager reference to clean them up via close() itself, so
            # without this their temp dirs / atexit handlers leak until
            # interpreter exit.
            self.close()
            raise

    def _register_repo(self, repo: "SkillRepository", origin: SkillOrigin) -> None:
        for skill in repo.get_skills():
            skill.set_resource_loader(
                lambda name=skill.name, r=repo: r.get_resources(name)
            )
            skill.set_origin(origin)
            previous = self._skills.get(skill.name)
            if previous is not None:
                logger.warning(
                    "Skill '%s' from %s overrides earlier registration from %s",
                    skill.name,
                    origin,
                    previous.origin if previous.origin is not None else "<unknown>",
                )
            self._skills[skill.name] = skill
            self._repos[skill.name] = repo

    def close(self) -> None:
        """Close every opened :class:`SkillRepository`, releasing any temp directory
        materialized for URL / classpath-zip / package sources. Idempotent.

        Iterates ``_opened_repos`` rather than ``_repos.values()``: duplicate skill
        names overwrite the earlier repo's reference in ``_repos``, but the displaced
        repo is still owned and must be closed. Dedup by identity in case the same
        repo contributes multiple skills.
        """
        seen: set[int] = set()
        for repo in self._opened_repos:
            if id(repo) in seen:
                continue
            seen.add(id(repo))
            with contextlib.suppress(Exception):
                repo.close()

    def __enter__(self) -> "SkillManager":
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

File: runtime/src/main/java/org/apache/flink/agents/runtime/skill/SkillManager.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.runtime.skill;

import org.apache.flink.agents.api.skills.SkillSourceSpec;
import org.apache.flink.agents.api.skills.Skills;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;

import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.IdentityHashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Loads and indexes all skills referenced by a {@link Skills} configuration.
 *
 * <p>Mirrors the Python {@code flink_agents.runtime.skill.skill_manager.SkillManager}.
 *
 * <p>Owned by {@code ResourceContextImpl}; closed via {@code ResourceCache.close()} on operator
 * close, which cascades to each repository's temp directory. Avoid constructing a {@code
 * SkillManager} outside that flow without {@code try-with-resources}.
 */
public class SkillManager implements AutoCloseable {

    private static final Logger LOG = LoggerFactory.getLogger(SkillManager.class);

    private final Skills config;
    private final ClassLoader classLoader;
    private final Map<String, AgentSkill> skills = new LinkedHashMap<>();
    private final Map<String, SkillRepository> repos = new HashMap<>();

    /**
     * Every {@link SkillRepository} this manager has opened, in load order. Kept separately from
     * {@link #repos} because that map is keyed by skill name — when two sources register the same
     * skill name, the second {@code put} drops the first repo's reference. Close logic iterates
     * this list (dedup'd by identity) to avoid leaking those displaced repos.
     */
    private final List<SkillRepository> openedRepos = new ArrayList<>();

    /**
     * Construct a {@code SkillManager} that resolves {@code classpath:} sources against {@code
     * classLoader}. Production code passes the Flink user-code class loader (threaded through
     * {@code ResourceCache} from {@code ActionExecutionOperator}); tests / standalone use may use
     * {@link #SkillManager(Skills)}.
     */
    public SkillManager(Skills config, ClassLoader classLoader) {
        this.config = config;
        this.classLoader = classLoader;
        loadAll();
    }

    /**
     * Convenience overload that uses {@link Thread#getContextClassLoader} for {@code classpath:}
     * resolution. Prefer {@link #SkillManager(Skills, ClassLoader)} in production paths where the
     * Flink user-code class loader is available — the thread context loader is not guaranteed to be
     * the user-code loader (e.g. Python interpreter or async-pool threads).
     */
    public SkillManager(Skills config) {
        this(config, Thread.currentThread().getContextClassLoader());
    }

    public int size() {
        return skills.size();
    }

    public AgentSkill getSkill(String name) {
        AgentSkill skill = skills.get(name);
        if (skill == null) {
            throw new IllegalArgumentException(
                    "Skill "
                            + name
                            + " not found, available skill names are: "
                            + getAllSkillNames());
        }
        return skill;
    }

    public List<String> getAllSkillNames() {
        return new ArrayList<>(skills.keySet());
    }

    @Nullable
    public String loadSkillResource(String skillName, String resourcePath) {
        return getSkill(skillName).getResource(resourcePath);
    }

    /** Build the {@code <available_skills>} system prompt for the given skill names. */
    public String generateDiscoveryPrompt(List<String> names) {
        if (size() == 0) {
            return "";
        }
        StringBuilder sb = new StringBuilder(SkillPromptProvider.SKILL_DISCOVERY_PROMPT);
        for (String name : names) {
            AgentSkill skill = getSkill(name);
            sb.append(
                    String.format(
                            SkillPromptProvider.AVAILABLE_SKILL_TEMPLATE,
                            skill.getName(),
                            skill.getDescription()));
        }
        sb.append(SkillPromptProvider.AVAILABLE_SKILLS_TAG_END);
        return sb.toString();
    }

    /**
     * Absolute directory paths for the listed skill names (filesystem-backed only). When called
     * with an empty or {@code null} list, returns directories for all filesystem-backed skills.
     */
    public List<String> getSkillDirs(List<String> names) {
        Iterable<String> selected = (names == null || names.isEmpty()) ? repos.keySet() : names;
        List<String> dirs = new ArrayList<>();
        for (String skillName : selected) {
            SkillRepository repo = repos.get(skillName);
            if (repo == null) {
                throw new IllegalArgumentException(
                        "Skill "
                                + skillName
                                + " not found, available skill names are: "
                                + new ArrayList<>(repos.keySet()));
            }
            Path dir = repo.getSkillDir(skillName);
            if (dir != null) {
                dirs.add(dir.toString());
            }
        }
        return dirs;
    }

    /** Return absolute directory path for a single skill, if filesystem-backed. */
    @Nullable
    public Path getSkillDir(String skillName) {
        SkillRepository repo = repos.get(skillName);
        return repo == null ? null : repo.getSkillDir(skillName);
    }

    private void loadAll() {
        try {
            for (SkillSourceSpec spec : config.getSources()) {
                try {
                    SkillRepository repo =
                            SkillSourceRegistry.get(spec.getScheme())
                                    .open(spec.getParams(), classLoader);
                    openedRepos.add(repo);
                    registerRepo(repo, originOf(spec));
                } catch (IOException | IllegalArgumentException e) {
                    throw new IllegalStateException(
                            "Failed to load skills from "
                                    + spec.getScheme()
                                    + ":"
                                    + spec.getParams(),
                            e);
                }
            }
        } catch (Throwable t) {
            // Release every repo opened so far, on any failure path. The caller never
            // receives a SkillManager reference (we're throwing from the constructor
            // path), so without this cleanup their shutdown hooks + temp dirs would leak
            // until JVM exit. The original failure propagates unchanged; a cleanup
            // failure rides along as suppressed so neither is lost — including an Error,
            // which closeRepos() does not catch per-repo and which would otherwise
            // replace the original failure outright.
            try {
                closeRepos();
            } catch (Throwable cleanupError) {
                t.addSuppressed(cleanupError);
            }
            throw t;
        }
    }

    /**
     * Build a {@link SkillOrigin} from a spec for diagnostics (WARN on duplicates, etc.). The
     * location description is delegated to the handler registered for the scheme — see {@link
     * SkillSourceHandler#describeLocation(Map)} — so adding a new scheme is a single registry call.
     */
    private static SkillOrigin originOf(SkillSourceSpec spec) {
        SkillSourceHandler handler = SkillSourceRegistry.get(spec.getScheme());
        return new SkillOrigin(spec.getScheme(), handler.describeLocation(spec.getParams()));
    }

    /**
     * Close every owned {@link SkillRepository}, releasing any temp directories materialized for
     * URL / classpath-zip / classpath-jar sources. Idempotent (delegated repos use {@link
     * java.util.concurrent.atomic.AtomicBoolean} guards).
     *
     * <p>Mirrors {@code ResourceCache.close()}: the first failure is rethrown after every repo has
     * been attempted, with subsequent failures attached as suppressed exceptions. This surfaces
     * real shutdown bugs (locked files, permission denied, disk full) instead of silently
     * swallowing them.
     */
    @Override
    public void close() throws Exception {
        closeRepos();
    }

    private void closeRepos() throws Exception {
        // Iterate openedRepos rather than repos.values(): duplicate skill names cause a later
        // registration to overwrite the earlier repo reference in `repos`, but the earlier repo
        // is still owned and must be closed. Dedup by identity in case the same repo
        // contributes multiple skills.
        Set<SkillRepository> unique = Collections.newSetFromMap(new IdentityHashMap<>());
        unique.addAll(openedRepos);
        Exception firstException = null;
        for (SkillRepository repo : unique) {
            try {
                repo.close();
            } catch (Exception e) {
                if (firstException == null) {
                    firstException = e;
                } else {
                    firstException.addSuppressed(e);
                }
            }
        }
        if (firstException != null) {
            throw firstException;
        }
    }

    private void registerRepo(SkillRepository repo, SkillOrigin origin) {
        for (AgentSkill skill : repo.getSkills()) {
            final String skillName = skill.getName();
            skill.setResourceLoader(() -> repo.getResources(skillName));
            skill.setOrigin(origin);
            AgentSkill previous = skills.put(skillName, skill);
            if (previous != null) {
                LOG.warn(
                        "Skill '{}' from {} overrides earlier registration from {}",
                        skillName,
                        origin,
                        previous.getOrigin() == null ? "<unknown>" : previous.getOrigin());
            }
            repos.put(skillName, repo);
        }
    }
}
